### ISSUE TITLE
Tracking issue to bootstrap `clang` with `clang`'s own `libc` on `aarch64`

### ISSUE THREAD

**Issue by izaakschroeder** [llvm/llvm-project#97191]
Tracking issue from discussions with @jhuber6 on the LLVM discord.

Known issues:

- [ ]`libc` + `aarch64` missing headers:
  - [ ] libc.include.sys_auxv
  - [ ] libc.include.fcntl
  - [ ] libc.include.sched
  - [ ] libc.include.sys_stat
- [ ]`libc` + `aarch64` missing entrypoints:
  - [ ] libc.src.assert.__assert_fail
  - [ ] libc.src.sys.auxv.getauxval

**Comment by boomanaiden154** [comment]
I was playing around with getting this setup with a current tip of tree checkout and came up with about the following (intended to be run in the CI container): https://gist.github.com/boomanaiden154/3750344f9660948648486be06f2ba5dd

There are a couple minor patches needed to get libunwind building (mostly because it expects `dladdr`) that we can probably hack away in the build with a preprocessor define. Other than that, everything works minus the actual clang build.

**Comment by jhuber6** [comment]
We *really* need to just provide an implementation of `malloc` that doesn't require special CMake flags. Either by pulling it from somewhere else or changing SCUDO to be a header.

I'm thinking the proper way to start handling this is to use something like this.
```
-DLLVM_RUNTIME_TARGETS=x86_64-unknown-linux-llvm
-DRUNTIMES_x86-unknown-linux-llvm_LLVM_ENABLE_RUNTIMES=compiler-rt;libc;libunwind;libcxxabi;libcxx
```
Doing `clang --target=x86-unknown-linux-llvm` should then ideally tell it to use the LLVM toolchain instead of the GNU one. That's much less destructive than shipping the `crt1.o` files so that they conflict with the system libc. I think we're only missing a few changes in the clang driver to make that work as expected, pretty much not linking the GCC startup libraries.

**Comment by boomanaiden154** [comment]
> We really need to just provide an implementation of malloc that doesn't require special CMake flags. Either by pulling it from somewhere else or changing SCUDO to be a header.

It should be pretty easy to just pull the sources into the main LLVM libc build and do everything there by default. If
```
cd compiler-rt/lib && /home/gha/sysroot/bin/clang++ \
  --sysroot=/home/gha/sysroot -fno-exceptions -fno-rtti \
  -std=c++17 -nostdinc++ -static -O3 -c \
  -I scudo/standalone/include \
  scudo/standalone/*.cpp
```
works, that can't be very hard. Although we do need to make sure we're building against the LLVM libc headers rather than the system libc headers.

**Comment by vonosmas** [comment]
I've recently ran a test on Linux-x86-64 to see where we are right now. This was done as a part of downstream Bazel-based configuration, so pasting the logs would not be particularly helpful. However, the list of issues observed is relatively small. I'll copy them below, and then update the bullets in the top comment above to track them:

* Full libc++ support is needed - I've ran the lbc++ build with wide-characters and threads already enabled, but missing filesystem support (see https://github.com/llvm/llvm-project/issues/191075#issuecomment-4392220136) means we can't construct `std::basic_ifstream<char>`
* llvm::thread should support non-integral `pthread_t` variants. We should use the same API / approach as libc++ which uses `pthread_id_np_t` - see 493ebf711e3f6641338e153ecdd284f262b4fb65
* Missing `struct winsize` and `TIOCGWINSZ` in `<sys/ioctl.h>` for https://github.com/llvm/llvm-project/blob/889496c13bd4d3114e375bd0aeb1f161eff67c33/llvm/lib/Support/Unix/Process.inc#L325
* Missing `pthread_setschedparam` for https://github.com/llvm/llvm-project/blob/889496c13bd4d3114e375bd0aeb1f161eff67c33/llvm/lib/Support/Unix/Threading.inc#L287
* Support for `_SC_ARG_MAX` in `sysconf` for https://github.com/llvm/llvm-project/blob/889496c13bd4d3114e375bd0aeb1f161eff67c33/llvm/lib/Support/Unix/Program.inc#L550
* Missing `<pwd.h>` and `getpwnam_r` for https://github.com/llvm/llvm-project/blob/889496c13bd4d3114e375bd0aeb1f161eff67c33/llvm/lib/Support/Unix/Path.inc#L715

The other issues on my end can be worked around by the using of feature macro:
* Missing `getrusage` and `RUSAGE_SELF` (gated by HAVE_GETRUSAGE) https://github.com/llvm/llvm-project/blob/889496c13bd4d3114e375bd0aeb1f161eff67c33/llvm/lib/Support/Unix/Process.inc#L53
* Missing `<execinfo.h>` (gated by HAVE_BACKTRACE)
* Missing `struct mallinfo` (gated by HAVE_MALLINFO)

**Comment by vonosmas** [comment]
The most recent round of testing that picks up merged updates to llvm-libc and enables unfinished endpoints (swprintf for wchar support in llvm-libc and realpath for filesystem support in llvm-libc), the only issues are:

* Missing <pwd.h> and `getpwnam_r`
* Missing `pthread_setschedparam` (under review in https://github.com/llvm/llvm-project/pull/205770)
* Missing definitions for `SA_NODEFER` and `SA_RESETHAND` for https://github.com/llvm/llvm-project/blob/57a7e7538cfc73184b8280160e8d89cedab04c5d/llvm/lib/Support/Unix/Signals.inc#L329
* `llvm::thread` should support non-integral pthread_t variants (under review in https://github.com/llvm/llvm-project/pull/207918)

As noted above, it's possible to disable at configuration time several other functionality that is currently missing: getrusage, mallinfo, backtrace, sbrk.

**Comment by vonosmas** [comment]
Several issues from comment https://github.com/llvm/llvm-project/issues/97191#issuecomment-4901605134 has been fixed. getrusage was added by cd6ceaf83146e52113e15ddf3392d3a22e96212e (and thus can be re-enabled). With backtrace/sbrk/mallinfo disabled I observe the following issues, in 

* `getpwnam_r` and `getpwuid_r` from `<pwd.h>` (and `_SC_GETPW_R_SIZE_MAX` value). I believe @kaladron has a WIP patches to address this (next in line - https://github.com/llvm/llvm-project/pull/214720)
* missing `<sys/vfs.h>` header on Linux (we'll just add a redirect to `<sys/statfs.h>` which is implemented).
* missing `struct flock` (it's a configuration issue, the type header exists, but it's just not exposed through `<fcntl.h>`
* missing `umask` from `<sys/stat.h>`
* `futimens` or `futimes` can be technically disabled by configuration, but would be nice to have.

**Comment by vonosmas** [comment]
After the recent updates, I see an issue with missing `umask` (tracked in https://github.com/llvm/llvm-project/issues/220759) which is used in Support/Path.inc: https://github.com/llvm/llvm-project/blob/2cbdbc9ad7ff40457023c8f353161069df8eb57f/llvm/lib/Support/Unix/Path.inc#L805-L811

and we still need to add a `_SC_GETPW_R_SIZE_MAX` value for sysconf.

**Comment by vonosmas** [comment]
After that it's the link step that fails due to the missing math library functions: `tanh`, `cosh`, `sinh`, `erf` that the ConstantFolding requires to be available on the host platform. I suspect these pop us link errors because compilation picks up libc++ `<cmath>` headers, which direct those calls to `__builtin_` variants which are in turn lowered to regular libcalls by Clang. @lntue FYI

**Comment by nickdesaulniers** [comment]
> tanh, cosh, sinh, erf that the ConstantFolding requires

I wonder if we get host-specific rounding differences due to those? 🤔

**Comment by vonosmas** [comment]
> > tanh, cosh, sinh, erf that the ConstantFolding requires
> 
> I wonder if we get host-specific rounding differences due to those? 🤔

Since ConstantFolding is executed on a host compiler with host libc -- definitely, unless the host libc has correct rounding.
